Update Experimental Features (#3077)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/experimental-features/index.md

## Allow extensions to load custom stylesheets
<!-- checkbox has a (?) link to this anchor wording -->

Some Microsoft Edge Add-ons can define custom color themes for DevTools. If you install an add-on that defines a custom color theme for DevTools, you need to enable the **Allow extensions to load custom stylesheets** experiment to view the add-on themes.

See also:
* [Apply a color theme to DevTools](https://learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/customize/theme)

Status:
* This checkbox is present in Microsoft Edge Canary 124.
* This checkbox is present in Microsoft Edge Stable 122.

<!-- ====================================================================== -->
## Capture node creation stacks
<!-- checkbox has a (?) link to this anchor wording -->

> To capture JavaScript stack traces when DOM nodes are added to the DOM at runtime, enable this experiment.  The captured stack traces are displayed in the **Stack Trace** pane of the **Elements** tool.<!-- todo: what tool contains this Stack Trace pane?  Sources tool?  Crash Analyzer tool?  Console?  where in that tool?  Update See Also accordingly -->

Author: What tool contains this Stack Trace pane? Sources tool? Crash Analyzer tool? Console? where in that tool?
@captainbrosset 
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/experimental-features/?branch=pr-en-us-3077#capture-node-creation-stacks

Reviewer: The **Stack Trace** pane is in the **Elements** tool, in the sidebar, next to **Styles**, **Computed**, etc.

![image](https://github.com/MicrosoftDocs/edge-developer/assets/1152698/a1ad530d-3bff-4520-b5b9-04c0ea4ed8e8)

After you enable the experiment, here's how to test:

1. Open the Elements tool
2. Use a webpage that creates DOM nodes dynamically with JS, such as the demo-to-do app
3. Create a new TODO item
4. Select the new TODO item with the inspect tool
5. Open the **Stack Trace** pane.

See also:
* [Configure the Shallow Size column to include an entire object's size](../memory-problems/heap-snapshots.md#configure-the-shallow-size-column-to-include-an-entire-objects-size) in _Record heap snapshots using the Memory tool_. 

Status:
* This checkbox is present in Microsoft Edge Canary 124.
* This checkbox is not present in Microsoft Edge Stable 122.

<!-- ====================================================================== -->
## Timeline: event initiators
<!-- checkbox has a (?) link to this anchor wording -->

Whether to include event initiators in the **Performance** tool.  This is useful to link a selected event to the other event that initiated it.

See also:
* [Introduction to the Performance tool](../evaluate-performance/index.md)

Status:
* This checkbox is not present in Microsoft Edge Canary 124.
* This checkbox is not present in Microsoft Edge Stable 122.

> The present section is retained because older browsers have a (?) link pointing to it.

Author: deleted section

* [Record heap snapshots using the Memory tool](../memory-problems/heap-snapshots.md)

Status:
* This checkbox is present in Microsoft Edge Canary 124.
* This checkbox is present in Microsoft Edge Stable 122.

<!-- ====================================================================== -->

> ## In heap snapshots, treat backing store size as part of the containing object

Author: review - @captainbrosset

See also:
* [Configure the Shallow Size column to include an entire object's size](../memory-problems/heap-snapshots.md#configure-the-shallow-size-column-to-include-an-entire-objects-size) in _Record heap snapshots using the Memory tool_. 

Status:
* This checkbox is present in Microsoft Edge Canary 124.
* This checkbox is not present in Microsoft Edge Stable 122.

<!-- ====================================================================== -->
## Timeline: event initiators
<!-- checkbox has a (?) link to this anchor wording -->

> Whether to include event initiators in the Timeline.<!-- todo: in the **Performance** tool and **Memory** tool. -->

Author: Perf tool, Mem tool, or both?  @captainbrosset

Author: ```
## Timeline: event initiators

Whether to include event initiators in the **Performance** tool.  This is useful to link a selected event to the other event that initiated it.

See also:
* [Introduction to the Performance tool](../evaluate-performance/index.md)
```

![The Sources tool, showing indentation markers](./index-images/indentation-markers.png)

See also:
* [Sources tool overview](../sources/index.md)

Status:
* This checkbox is present in Microsoft Edge Canary 124.
* This checkbox is present in Microsoft Edge Stable 122.

<!-- ====================================================================== -->
## WebAssembly Debugging: Enable DWARF support
<!-- checkbox has a (?) link to this anchor wording -->

Enables DWARF support for WebAssembly debugging.
<!-- todo old: need text -->

See also:
* [Improved WebAssembly debugging](../whats-new/2019/12/devtools.md#improved-webassembly-debugging) in _What's New in DevTools (Microsoft Edge 80)_.

Status:
* This checkbox is not present in Microsoft Edge Canary 124.
* This checkbox is not present in Microsoft Edge Stable 122.

> The present section is retained because older browsers have a (?) link pointing to it.

Author: deleted section

Enables experimental cookie features.
<!-- todo old: need text -->

<!-- todo: link
See also:
* []() -->

Status:
* This checkbox is present in Microsoft Edge Canary 124.
* This checkbox is present in Microsoft Edge Stable 122.

<!-- ====================================================================== -->

> ## Deprecate CSS \<length\> authoring tool in the Styles tab

Author: removed entry

Enables experimental cookie features.
<!-- todo old: need text -->

See also:
* [View, edit, and delete cookies](../storage/cookies.md)

Status:
* This checkbox is present in Microsoft Edge Canary 124.
* This checkbox is present in Microsoft Edge Stable 122.

<!-- ====================================================================== -->
## Display more precise changes in the Changes tab
<!-- checkbox has a (?) link to this anchor wording -->

Whether to display changes by using a more detailed layout in the **Changes** tool.

See also:
* [More precise changes in the Changes tab](https://developer.chrome.com/blog/new-in-devtools-98/#changes)

Status:
* This checkbox is not present in Microsoft Edge Canary 124.
* This checkbox is not present in Microsoft Edge Stable 122.

> The present section is retained because older browsers have a (?) link pointing to it.

Author: deleted section

Whether to use local overrides for response headers.
<!-- todo old: need text -->

See also:
* [Override network response headers](https://developer.chrome.com/blog/new-in-devtools-113/#network)
* [Add columns for response headers](../network/reference.md#add-columns-for-response-headers) in _Network features reference_.

Status:
* This checkbox is not present in Microsoft Edge Canary 124.
* This checkbox is not present in Microsoft Edge Stable 122.

> The present section is retained because older browsers have a (?) link pointing to it.

Author: deleted section

* [Change colors with the Color Picker](../css/reference.md#change-colors-with-the-color-picker) in _CSS features reference_

Status:
* This checkbox is not present in Microsoft Edge Canary 124.
* This checkbox is not present in Microsoft Edge Stable 122.

> The present section will be removed, because current browsers no longer have this checkbox, and older browsers don't have a (?) link pointing to this section.

Author: deleted section

* This checkbox is not present in Microsoft Edge Canary 125.
* This checkbox is present in Microsoft Edge Stable 123.

<!-- ====================================================================== -->
## Autofill panel
<!-- no (?) link -->
<!-- got moved down -->

_new as of Microsoft Edge 125_

Adds an **Autofill** tool to DevTools, to record various events that the browser triggers when a user is filling in a form by using the browser's autofill feature.  For example, when the browser remembers your address and autofills it in an address field.

![The Autofill tool](./index-images/autofill-tool.png)

See also:
* [Autofill](../../privacy-whitepaper/index.md#autofill) in _Microsoft Edge Privacy Whitepaper_.
* [Autofill](../../webview2/concepts/overview-features-apis.md?tabs=dotnetcsharp#autofill) in _Overview of WebView2 features and APIs_.

Status:
* This checkbox is present in Microsoft Edge Canary 125.
* This checkbox is present in Microsoft Edge Stable 123, as **Enable Autofill view**.

<!-- ====================================================================== -->
## Timeline: Show postMessage dispatch and handling flows
<!-- no (?) link -->

_new as of Microsoft Edge 124_

> Distinguishes postMessage events from other timeline events, to easily investigate performance issues related to post messages across various threads in applications.

Reviewer: ```suggestion
Distinguishes events that are triggered by the `postMessage` method from other events displayed in the **Performance** tool, to easily investigate performance issues that are related to postin messages across various threads of an application.
```

Author: todo before Resolve: update What's New PR per this rewording.
```
Distinguishes events that are triggered by the `postMessage` method from other events that are displayed in the **Performance** tool, to easily investigate performance issues that are related to posting messages across various threads of an application.
```

> This experiment improves the **Performance** tool's timeline to help you quickly identify `postMessage` events and handlers, by distinguishing `postMessage` events from other timeline events.  Without this experiment, `postMessage` dispatch events and handler events both appear as generic scripting function call events.  With this experiment enabled:

Reviewer: ```suggestion
This experiment improves the **Performance** tool's **Main** section to help you quickly identify `postMessage` events and handlers, by distinguishing `postMessage` events from other events.  Without this experiment, events triggered by dispatching and handling messages between threads of an application appear as generic scripting function call events.  With this experiment enabled:
```

Author: todo: before Resolve, update What's New PR:
```
This experiment improves the **Performance** tool's **Main** section to help you quickly identify `postMessage` events and handlers, by distinguishing `postMessage` events from other events.  Without this experiment, events that are triggered by dispatching and handling messages between threads of an application appear as generic scripting function-call events.  With this experiment enabled:
```

* [Performance recording event reference](../evaluate-performance/performance-reference.md)
* [postMessage Trace Events demo (rendered)](https://microsoftedge.github.io/Demos/devtools-postmessage-perf-timeline/)
* [postMessage Trace Events demo (source code)](https://github.com/MicrosoftEdge/Demos/tree/main/devtools-postmessage-perf-timeline)
<!--
* [postMessage events are distinguished from other timeline events](https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/whats-new/2024/04/devtools-124?branch=pr-en-us-3129#postmessage-events-are-distinguished-from-other-timeline-events) in _What's New in DevTools (Microsoft Edge 124)_.
* [View messages between windows, iframes, and dedicated workers] in _Performance features reference_.
(in open PR 3132)
-->

Status:
* This checkbox is present in Microsoft Edge Canary 125.
* This checkbox is not present in Microsoft Edge Stable 123.

<!-- ====================================================================== -->
## Enable save and load trace with annotations in Performance Panel
<!-- no (?) link -->

_new as of Microsoft Edge 125_

> Enables save and load trace with annotations in the **Performance** tool.  Adds the ability to persist the following two items inside trace files, and reapply them when importing traces that contain breadcrumbs and altered tracks.

Reviewer: ```suggestion
Adds the ability to persist the following changes made to trace files in the **Performance** tool, and reapply them when importing traces:
```

> * Minimap-driven navigation, also called "breadcrumbs".  This feature enables you to crop into and focus on the more relevant parts of a trace.  See [Breadcrumbs in the Timeline](https://developer.chrome.com/blog/new-in-devtools-122?hl=en#perf-breadcrumbs) in _What's new in DevTools (Chrome 122)_.

Reviewer: ```suggestion
* Changes made by using the breadcrumbs. This feature enables you to crop into and focus on the more relevant parts of a trace.  See [Breadcrumbs in the Timeline](https://developer.chrome.com/blog/new-in-devtools-122?hl=en#perf-breadcrumbs) in _What's new in DevTools (Chrome 122)_.
```

Author: ```
* Changes made by using the breadcrumbs.  This feature enables you to crop into and focus on the more relevant parts of a trace.  See [Breadcrumbs in the Timeline](https://developer.chrome.com/blog/new-in-devtools-122#perf-breadcrumbs) in _What's new in DevTools (Chrome 122)_.
```
trimmed url

> * Track data context-menu.  This feature enables you to merge and hide superfluous or irrelevant entries of a trace, cleaning up the flame charts and trees.  See [Hide functions and their children in the flame chart](https://developer.chrome.com/blog/new-in-devtools-124#hide-func) in _What's new in DevTools (Chrome 124)_.

Reviewer: ```suggestion
* Changes made by using the track contextual menu.  This feature enables you to merge and hide superfluous or irrelevant entries of a trace, cleaning up the flame charts and trees.  See [Hide functions and their children in the flame chart](https://developer.chrome.com/blog/new-in-devtools-124#hide-func) in _What's new in DevTools (Chrome 124)_.
```

See also:
* [Timeline event reference](/microsoft-edge/devtools-guide-chromium/evaluate-performance/performance-reference)
* [PostMessage Trace Events demo (rendered)](https://microsoftedge.github.io/Demos/devtools-postmessage-perf-timeline/)
* [PostMessage Trace Events demo (source code)](https://github.com/MicrosoftEdge/Demos/tree/main/devtools-postmessage-perf-timeline)
<!-- * [postMessage events are distinguished from other timeline events](https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/whats-new/2024/04/devtools-124?branch=pr-en-us-3129#postmessage-events-are-distinguished-from-other-timeline-events) in _What's New in DevTools (Microsoft Edge 124)_. -->

Status:
* This checkbox is present in Microsoft Edge Canary 125.
* This checkbox is not present in Microsoft Edge Stable 123.

<!-- ====================================================================== -->
## Enable save and load trace with annotations in Performance Panel
<!-- no (?) link -->

_new as of Microsoft Edge 125_

Enables save and load trace with annotations in the **Performance** tool.

> <!-- todo: need desc -->

Author: Enable save and load trace with annotations in Performance Panel - need real description, if possible.
@captainbrosset

Reviewer: I couldn't find anything about this. I will ask Rob to help if he can. The bug that introduced this in Chromium was done by the Chrome team, and is unfortunately not publicly visible.

Reviewer: Trying to find out more about this experiment. Will report when I know more.

Author: The "minimap-driven navigation" (aka "breadcrumbs", a feature that enables developers to crop into and focus on the more relevant parts of a trace, see WNIDT 122:
https://developer.chrome.com/blog/new-in-devtools-122?hl=en#perf-breadcrumbs)
and the "track data context-menu" (a feature that enables developers to merge and hide superfluous or irrelevant entries of a trace, cleaning up the flame charts and trees, see WNIDT 124:
https://developer.chrome.com/blog/new-in-devtools-124#hide-func)
This experiment added the ability to persist those inside trace files, and reapply them when importing traces that contain breadcrumbs and altered tracks.

from @captainbrosset offline

*  Progressive Web Apps (PWAs)
*  Other common web development issues

The [webhint](https://webhint.io) experiment displays webhint feedback in the **Issues** tool.  Select an issue to display documentation about the solution and a list of the affected resources on your website.  Select a resource link to open the relevant tool, such as **Network**, **Sources**, or **Elements**:

![webhint feedback in the Issues tool](./index-images/experiments-webhint.png)

See also:
* [Find and fix problems using the Issues tool](../issues/index.md)

Status:
* This checkbox is present in Microsoft Edge Canary 125.
* This checkbox is present in Microsoft Edge Stable 123.

<!-- ====================================================================== -->
## Show issues in Elements
<!-- checkbox has a (?) link to this anchor wording -->

> Displays syntax errors under HTML in the **DOM** view of the **Elements** tool.

Reviewer: ```suggestion
Displays syntax errors as wavy underlines under DOM nodes in the **Elements** tool.
```

Author: replaced

When DevTools is used to debug processes that are running in Node.js, this experiment makes whatever `console.profile` generates available to the **Performance** Tool.

See also:
* [JavaScript Profiler tool](../javascript-profiler/javascript-profiler-tool.md)

Status:
* This checkbox is not present in Microsoft Edge Canary 125.
* This checkbox is present in Microsoft Edge Stable 123.

<!-- ====================================================================== -->

> ## Timeline: Enable track confirmation feature that can reorder or hide a track in the flame chart

Reviewer: ```suggestion
## Timeline: Enable track configuration feature that can reorder or hide a track in the flame chart
```

Author: confirmed in UI (d/k why grayed out, i think we discussed or logged bug):
![image](https://github.com/MicrosoftDocs/edge-developer/assets/45407486/842ed439-ed5a-4cbb-91de-cc481af23a66)

<!-- no (?) link -->

_new as of Microsoft Edge 125_

> Enables the track confirmation feature, which can reorder or hide a track in the flame chart.

Reviewer: ```suggestion
Enables the track configuration feature, which lets you choose which tracks are visible in the **Performance** tool, and the order in which they appear. After you enable this experiment, to configure tracks:

1. Create a new performance recording in the **Performance** tool.
1. Hover over the tracks in the **Performance** tool, such as **Network**, **Interactions**, or **Main**. The track configuration icon (pen icon) appears on the left of the hovered track.
1. Click the track configuration icon. The track configuration screen appears.
1. On the track configuration screen, click the up or down arrows to change the order of the tracks, or click on the eye icons to toggle tracks.
1. To confirm the new track configuration, click the checkmark icon to the right of a track. The performance recording is displayed, with the new track configuration.
```

Author: revised draft, will check against UI:
```
Enables the track configuration feature, which lets you choose which tracks are visible in the **Performance** tool, and the order in which they appear.  After you enable this experiment, to configure tracks:
1. In the **Performance** tool, create a new performance recording.
1. In the **Performance** tool, hover over a track, such as **Network**, **Interactions**, or **Main**.  The **Track configuration** (pen icon) button appears on the left of the hovered track.
1. Click the **Track configuration** (pen icon) button.  The **track configuration** screen opens.
1. On the **track configuration** screen, click the up or down arrows to change the order of the tracks, or click on the eye icons to toggle tracks.
1. To confirm the new track configuration, click the checkmark icon to the right of a track.  The performance recording is displayed, with the new track configuration.

See also:
* [Introduction to the Performance tool](../evaluate-performance/index.md)
```

> <!-- todo: need desc 

Author: Timeline: Enable track confirmation feature that can reorder or hide a track in the flame chart
Need real description, if possible.  What tool is this about?
@captainbrosset

Reviewer: See comment above.

add: in the _Foo_ tool.
-->

<!--
See also:
* []()
-->

Status:
* This checkbox is present in Microsoft Edge Canary 125.
* This checkbox is not present in Microsoft Edge Stable 123.

<!-- ====================================================================== -->
## Timeline: Enable User Timings based extensions
<!-- no (?) link -->

_new as of Microsoft Edge 125_

Enables extensions that are based on user timings.

> <!-- todo: need desc

Author: Timeline: Enable User Timings based extensions
Need real description, if possible.  What tool is this about?
@captainbrosset

Reviewer: The upstream bug that introduced this experiment is also private. So I wasn't able to get much information about this.
My current understanding is this: users can install add-ons in the browser that add new tracks to the list of tracks displayed in the Performance tool. One such add-on could add a new track that displays a series of events over time, just like the Main or Interactions tracks do, but would show other types of events.

Reviewer: Trying to find out more about this experiment. Will report when I know more.

Author: Enables extensions that customize the appearance of user timing events in the Performance tool.
from @captainbrosset offline
